jenkins-master/jobs/scripts/workspace/submission.py

```
@@ -0,0 +1,400 @@
+        if nargs is not None:
+            raise ValueError("nargs not allowed")
+        argparse.Action.__init__(self, option_strings, dest, **kwargs)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        if type(values) is str:
+            values = values if values != 'None' else None
+        elif type(values) is list:
+            values = [value if value != 'None' else None for value in values]
+
+        setattr(namespace, self.dest, values)
+
+
+class FirefoxUIResultParser(object):
+
+    BUSTED = 'busted'
+    SUCCESS = 'success'
+    TESTFAILED = 'testfailed'
+    UNKNOWN = 'unknown'
+
+    def __init__(self, retval, log_file):
+        self.retval = retval
+        self.log_file = log_file
+        self.failure_re = re.compile(r'(^TEST-UNEXPECTED-FAIL|TEST-UNEXPECTED-ERROR)|'
+                                     r'(.*CRASH: )|'
+                                     r'(Crash reason: )')
+        self.failures = []
+        self.parse()
+
+    def parse(self):
+        try:
+            with open(self.log_file, 'r') as f:
+                for line in f.readlines():
+                    if self.failure_re.match(line):
+                        self.failures.append(line)
+        except IOError:
+            print('Cannot parse failures due to missing log file: {}'.format(self.log_file))
+
+    @property
+    def status(self):
+        status = self.UNKNOWN
+
+        # retval.txt was not written - so most likely an abort
+        if self.retval is None or (self.retval and not self.failures):
+            status = self.BUSTED
+
+        elif not self.failures:
+            status = self.SUCCESS
+
+        elif self.failures:
+            status = self.TESTFAILED
+
+        return status
+
+    def failures_as_json(self):
+        failures = {'all_errors': [], 'errors_truncated': True}
+
+        for failure in self.failures:
+            failures['all_errors'].append({'line': failure, 'linenumber': None})
```

`'linenumber': None` prevent the failure summary from being displayed in the Treeherder log viewer. You can use `1` for all failures instead of `None` to get around this for now.